hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java

```diff
@@ -0,0 +1,178 @@
+        }
+
+        SqlPage page = fetchInternal(clientCursor, cursorBufferSize, serializationService);
+
+        if (page.isLast()) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return page;
+    }
+
+    private SqlPage fetchInternal(
+        QueryClientState clientCursor,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        Iterator<SqlRow> iterator = clientCursor.getIterator();
+
+        List<SqlPageRow> page = new ArrayList<>(cursorBufferSize);
+        boolean last = fetchPage(iterator, page, cursorBufferSize, serializationService);
+
+        if (last) {
+            deleteClientCursor(clientCursor);
+        }
+
+        return new SqlPage(page, last);
+    }
+
+    private static boolean fetchPage(
+        Iterator<SqlRow> iterator,
+        List<SqlPageRow> page,
+        int cursorBufferSize,
+        InternalSerializationService serializationService
+    ) {
+        while (iterator.hasNext()) {
+            SqlRow row = iterator.next();
+            SqlPageRow convertedRow = convertRow(row, serializationService);
+
+            page.add(convertedRow);
+
+            if (page.size() == cursorBufferSize) {
+                break;
+            }
+        }
+
+        return !iterator.hasNext();
+    }
+
+    private static SqlPageRow convertRow(SqlRow row, InternalSerializationService serializationService) {
+        int columnCount = row.getMetadata().getColumnCount();
+
+        List<Data> values = new ArrayList<>(columnCount);
+
+        for (int i = 0; i < columnCount; i++) {
+            values.add(serializationService.toData(row.getObject(i)));
+        }
+
+        return new SqlPageRow(values);
+    }
+
+    public void close(UUID clientId, QueryId queryId) {
+        QueryClientState clientCursor = getClientCursor(clientId, queryId);
+
+        if (clientCursor != null) {
+            clientCursor.getSqlResult().close();
+
+            deleteClientCursor(clientCursor);
+        }
+    }
+
+    public void reset() {
+        clientCursors.clear();
+    }
+
+    public void update(Set<UUID> activeClientIds) {
+        List<QueryClientState> victims = new ArrayList<>();
+
+        for (QueryClientState clientCursor : clientCursors.values()) {
+            if (!activeClientIds.contains(clientCursor.getClientId())) {
+                victims.add(clientCursor);
+            }
+        }
+
+        for (QueryClientState victim : victims) {
+            QueryException error = QueryException.clientMemberConnection(victim.getClientId());
+
+            victim.getSqlResult().closeOnError(error);
+
+            deleteClientCursor(victim);
+        }
+    }
+
+    private QueryClientState getClientCursor(UUID clientId, QueryId queryId) {
+        QueryClientState cursor = clientCursors.get(queryId);
+
+        if (cursor == null || (clientId != null && !cursor.getClientId().equals(clientId))) {
```

Comment:
@devozerov This part is confusing to me. Can you document when can a client be null? To me, it looked like it can not, or maybe I am missing something.

Reply:
Fixed. Part of this expression was redundant.